README.md
# perftest-toolkit

[![Docker Repository on Quay](https://quay.io/repository/3scale/perftest-toolkit/status "Docker Repository on Quay")](https://quay.io/repository/3scale/perftest-toolkit)

This repo has tools and deployment configs for a performance testing environment to be able to run performance tests of a 3scale API Management solution, focussing on the traffic intensive parts of the solution (the API Gateway and the Service Management API).

We have open sourced it to enable partners, customers and support engineers to run their own performance tests on "self-managed" (i.e. Not SaaS) installations of the 3scale API Management solution.

By running performance test with the same tools, scripts, traffic patterns and measurements as we at 3scale do, we hope it will help produce results that can be more easily compared with the results we achieve in our regular in-house performance testing and that we can run internally.

The goal is to help to resolve doubts or issues related to scalability or performance more quickly and easily - allowing you to achieve the high levels of low-latency performance we strive for and ensure in our own internal testing.

## Table of Contents

* [High level overview](#high-level-overview)
* [Deploy injector](#deploy-injector)
   * [Common settings](#common-settings)
   * [Test your 3scale services](#test-your-3scale-services)
   * [Setup traffic profiles](#setup-traffic-profiles)
* [Run tests](#run-tests)
* [Sustained load](#sustained-load)
* [Troubleshooting](#troubleshooting)
   * [Check apicast gateway configuration](#check-apicast-gateway-configuration)
   * [Check backend listener traffic](#check-backend-listener-traffic)
   * [Check upstream service traffic](#check-upstream-service-traffic)

Generated using [github-markdown-toc](https://github.com/ekalinin/github-markdown-toc)

## High level overview

High level overview is quite simple. Main components are represented in the diagram below.

* Injector: Source of HTTP/HTTPS traffic requests
* Openshift Container Platform with 3scale installed
* Upstream API: Also named as backend API, this is the final API service as an HTTP traffic endpoint. Optionally, for testing purposes, [deploy Upstream API](deployment/doc/deploy-upstream-api.md).
* Test Configurator (Buddhi): 3scale setup and traffic generation tool

![Test setup](deployment/doc/infrastructure.png "Infrastructure")

## Deploy injector

There are **two** ways of running your tests and the injector has to be configured accordingly.

* [Test your own 3scale services](#test-your-3scale-services): The injector will be custom configured to use your 3scale products (a.k.a. services).

* [Setup traffic profiles](#setup-traffic-profiles): Configure your performance tests to use synthetically generated traffic based on traffic models.

**Requirements**:

Control node:
* ansible >= 2.9.14
* python >= 3.0
* Install ansible requirements
```
cd deployment
ansible-galaxy install -r requirements.yaml
```

Managed node host:
* Docker >= 1.12
* python >= 2.7
* docker-py >= 1.7.0

Make sure that the injector host’s hardware resources is not the performance tests bottleneck. Enough cpu, memory and network resources should be available.

### Common settings

** HTTP/HTTPS **

By default, the injector will generate HTTPS traffic on the port number 443.
You can change this setting editing *injector_hyperfoil_target_protocol* and *injector_hyperfoil_target_port* parameters.

```
File: group_vars/all.yml

injector_hyperfoil_target_protocol: https
injector_hyperfoil_target_port: 443
```

### Test your 3scale services

**Steps**:

**1.** Provide **hyperfoil controller** host IP address or DNS name and at least one **hyperfoil agent** host IP address or DNS name. For example:

```
File: hosts

[hyperfoil_controller]
controllerhost.example.com ansible_host=controllerhost.example.com ansible_role=root

[hyperfoil_agent]
agenthost.example.com ansible_host=agenthost.example.com ansible_role=root
```

**Note 01**: make sure defined ansible user has ssh login access to the host without password.

**Note 02**: make sure hyperfoil controller host has ssh login access to the agent host without password.

More than one hyperfoil agent can be configured. Useful when the injector becomes a bottleneck. For example to configure two agents:

```
File: hosts

[hyperfoil_controller]
controllerhost.example.com ansible_host=controllerhost.example.com ansible_role=root

[hyperfoil_agent]
agenthost01.example.com ansible_host=agenthost01.example.com ansible_role=root
agenthost02.example.com ansible_host=agenthost02.example.com ansible_role=root
```

**2.** Configure the following settings in `roles/user-traffic-reader/defaults/main.yml` file:
* `threescale_portal_endpoint`: 3scale portal endpoint
* `threescale_services`: Select the 3scale services you want to use for the tests. Leave it empty to use them all.

```
File: roles/user-traffic-reader/defaults/main.yml

# URI that includes your password and portal endpoint in the following format: <schema>://<password>@<admin-portal-domain>.
# The <password> can be either the provider key or an access token for the 3scale Account Management API.
# <admin-portal-domain> is the URL used to log into the admin portal.
# Example: https://access-token@account-admin.3scale.net
threescale_portal_endpoint: <THREESCALE_PORTAL_ENDPOINT>

# Comma separated list of services (Id's or system names)
# If empty, all available services will be used
threescale_services: ""
```

**3.** Execute the playbook `injector.yml` to deploy injector.

Avoid ssh issues when running anisble playbooks

```bash
$ cat ~/.ansible.cfg
[ssh_connection]
ssh_args = -o ServerAliveInterval=30
pipelining = True
```

Start the playbook

```bash
cd deployment/
ansible-playbook -i hosts injector.yml
```

### Setup traffic profiles

Skip these steps if testing your own 3scale services. This steps will setup 3scale services for performance testing.

**Steps**:

**1.** Provide **hyperfoil controller** host IP address or DNS name and at least one **hyperfoil agent** host IP address or DNS name. For example:

```
File: hosts

[hyperfoil_controller]
controllerhost.example.com ansible_host=controllerhost.example.com ansible_role=root

[hyperfoil_agent]
agenthost.example.com ansible_host=agenthost.example.com ansible_role=root
```

**Note 01**: make sure defined ansible user has ssh login access to the host without password.
**Note 02**: make sure hyperfoil controller host has ssh login access to the agent host without password.

More than one hyperfoil agent can be configured. Useful when the injector becomes a bottleneck. For example to configure two agents:

```
File: hosts

[hyperfoil_controller]
controllerhost.example.com ansible_host=controllerhost.example.com ansible_role=root

[hyperfoil_agent]
agenthost01.example.com ansible_host=agenthost01.example.com ansible_role=root
agenthost02.example.com ansible_host=agenthost02.example.com ansible_role=root
```

**2.** Configure the following settings in `roles/profiled-traffic-generator/defaults/main.yml` file:
* `threescale_portal_endpoint`: 3scale portal endpoint
* `traffic_profile`: Currently [available profiles](buddhi/README.md#profiles): `simple, backend, medium, standard`
* `private_base_url`: Private Base URL used for the tests. Make sure your private application behaves like an echo api service.
* `public_base_url`: Optionally, configure the `Public Base URL` used for the tests for self managed apicast environments. Otherwise, leave it empty.

```
File: roles/traffic-configurator/defaults/main.yml

# URI that includes your password and portal endpoint in the following format: <schema>://<password>@<admin-portal-domain>.
# The <password> can be either the provider key or an access token for the 3scale Account Management API.
# <admin-portal-domain> is the URL used to log into the admin portal.
# Example: https://access-token@account-admin.3scale.net
threescale_portal_endpoint: <THREESCALE_PORTAL_ENDPOINT>

# Used traffic for performance testing is not real traffic.
# It is synthetically generated traffic based on traffic models.
# Information about available traffic profiles (or test plans) can be found here:
# https://github.com/3scale/perftest-toolkit/blob/master/buddhi/README.md#profiles
# Currently available profiles: [ simple | backend | medium | standard ]
traffic_profile: <TRAFFIC_PROFILE>

# Private Base URL
# Make sure your private application behaves like an echo api service
# example: https://echo-api.3scale.net:443
private_base_url: <PRIVATE_BASE_URL>

# Public Base URL
# Public address of your API gateway in the production environment.
# Optional. When it is left empty, public base url will be the hosted gateway url
# example: https://gw.example.com:443
public_base_url: <PUBLIC_BASE_URL>
```

**3.** Execute the playbook `profiled-injector.yml` to deploy injector.

Avoid ssh issues when running anisble playbooks

```bash
$ cat ~/.ansible.cfg
[ssh_connection]
ssh_args = -o ServerAliveInterval=30
pipelining = True
```

Start the playbook

```bash
ansible-playbook -i hosts profiled-injector.yml

```

## Run tests

**1.** Configure testing settings:

```
File: run.yml

USERS_PER_SEC: Requests per second
DURATION_SEC: Duration of the performance test in seconds
SHARED_CONNECTIONS: Number of connections open per target HOST
```

**2.** Run tests

```bash
ansible-playbook -i hosts -i benchmarks/3scale.csv run.yml
```

**3.** View Report

The test results of the last execution are automatically stored in **deployment/benchmarks/index-<runid>.html**.
The html file can be directly opened with your favorite web browser.

## Sustained load

Some performance test are looking for *peak* and *sustained* traffic maximum performance.
*Sustained* traffic is defined as traffic load where *Job Queue* size is always at low levels, or even empty.
For *sustained* traffic performance benchmark, *Job Queue* must be monitorized.

This is a small guideline to monitor *Job Queue* size:

- Get backend redis pod

```bash
$ oc get pods | grep redis
backend-redis-2-nkrkk         1/1       Running   0          14d
```

- Get Job Queue size

```bash
$ oc rsh backend-redis-2-nkrkk /bin/sh -i -c 'redis-cli -n 1 llen resque:queue:priority'
(integer) 0
```

## Troubleshooting

Sometimes, even though all deployment commands run successfully, performance traffic may be broken.
This might be due to a misconfiguration in any stage of the deployment process.
When performance HTTP traffic response codes are not as expected, i.e. **200 OK**,
there are few checks that can be very handy to find out configuration mistakes.

### Check apicast gateway configuration

First, scale down *apicast-production* service to just one pod.

Monitor pod's logs for traffic accesslog.

```bash
oc logs -f apicast-production-X-podId
```

[Run tests](#run-tests) and check for logs.

Check response codes on accesslog.

If accesslog shows *could not find service for host* error, then the configured virtual hosts do not match traffic *Host* header. For example:
```
2018/06/05 13:32:41 [warn] 25#25: *883 [lua] errors.lua:43: get_upstream(): could not find service for host: 9ccd143c-dbe4-471c-9bce-41df7dde8d99.benchmark.perftest.3sca.net, client: 10.130.4.1, server: _, request: "GET /855aaf5c-a199-4145-a3ab-ea9402cc35db/some-request?user_key=32313d20d99780a5 HTTP/1.1", host: "9ccd143c-dbe4-471c-9bce-41df7dde8d99.benchmark.perftest.3sca.net"
```
Another issue might be when response codes are *404 Not Found*.Then proxy-rules do not match traffic path.

In anyone of the previous cases, it seems that *apicast gateway* does not have latest configuration.
Pods restart is required or wait until process fetches new configuration based on
*APICAST_CONFIGURATION_CACHE* apicast configuration parameter.

Restart is easily done downscaling to 0 and then scaling back to desired number of pods.

```bash
$ oc scale dc apicast-production --replicas=0
$ oc scale dc apicast-production --replicas=2
```

### Check backend listener traffic

First, scale down *backend-listener* service to just one pod.

Monitor pod's logs for traffic accesslog.

```bash
oc logs -f backend-listener-X-podId
```

[Run tests](#run-tests) and check for logs.

If no logs are shown, check [gateway troubleshooting section](#check-apicast-gateway-configuration)

If logs are shown, check response codes on accesslog. Other than *200 OK* means
- *redis* is down,
- *redis* address is misconfigured in *backend-listener*
- redis does not have required data to authenticate requests

### Check upstream service traffic

When *backend-listener* accesslog shows requests are being answered with *200 OK* response codes,
the last usual suspect is upstream or upstream configuration.

Check *upstream* uri is correctly configured in your 3scale configuration



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.4.0] - 2020-10-15

### Added

- Added Medium profile [PR #72](https://github.com/3scale/perftest-toolkit/pull/72)
- Hyperfoil 0.15.0 [PR #34](https://github.com/3scale/perftest-toolkit/pull/34)
- Traffic profiles with self-managed deployment option [PR #67](https://github.com/3scale/perftest-toolkit/pull/67)

## [2.2.2] - 2020-10-15

### Fixed

- fix: old 3scale releases returned 404 Not found on backend endpoints [PR #55](https://github.com/3scale/perftest-toolkit/pull/55)

## [2.2.1] - 2020-06-02

### Fixed

- fix standard concurrency [PR #49](https://github.com/3scale/perftest-toolkit/pull/49)

## [2.2.0] - 2020-06-02

### Added

- Provision standard profile concurrently [PR #46](https://github.com/3scale/perftest-toolkit/pull/46)

## [2.1.0] - 2020-05-26

### Added

- Added Simple and Backend profile [PR #36](https://github.com/3scale/perftest-toolkit/pull/36)
- Added Standard profile [PR #40](https://github.com/3scale/perftest-toolkit/pull/40)

## [2.0.0] - 2020-04-03

### Added

- Testing setup fetched from 3scale API [PR #31](https://github.com/3scale/perftest-toolkit/pull/31)
- Jmeter upgraded to 5.2 [PR #31](https://github.com/3scale/perftest-toolkit/pull/31)

### Fixes

- [Deployment] do not require interactive mode for 'docker run' command [PR #24](https://github.com/3scale/perftest-toolkit/pull/24)

## [1.1.0] - 2018-06-07

### Added

- [Buddhi] metric reporter service api [PR #6](https://github.com/3scale/perftest-toolkit/pull/6)


buddhi/lib/amp/toolkit/3scale/helper.rb
module AMP
  module Toolkit
    module ThreeScale
      module Helper
        def self.random_lowercase_name
          [*('a'..'z')].sample(8).join
        end

        def self.parse_uri(uri)
          # raises error when remote_str is not string, but object or something else.
          uri_obj = URI(uri)
          # URI::HTTP is parent of URI::HTTPS
          # with single check both types are checked
          raise "invalid url: #{uri}" unless uri_obj.kind_of?(URI::HTTP)

          uri_obj
        end

        def self.create_service(client, public_base_url)
          service_name = "PERF_TEST_#{random_lowercase_name}"
          system_name = service_name.delete("\s").downcase
          deployment_option = if public_base_url.nil?
                                'hosted'
                              else
                                'self_managed'
                              end
          svc_params = { 'name' => service_name, 'system_name' => system_name,
                         'deployment_option' => deployment_option }
          svc_obj = client.create_service svc_params

          if (errors = svc_obj['errors'])
            raise "Service has not been created: #{errors}"
          end

          svc_obj
        end

        def self.create_application_plan(client, service)
          name = random_lowercase_name
          plan_params = {
            'name' => name, 'state' => 'published', 'default' => false,
            'custom' => false, 'system_name' => name
          }

          plan_obj = client.create_application_plan service.fetch('id'), plan_params
          if (errors = plan_obj['errors'])
            raise "Application plan has not been created: #{errors}"
          end

          plan_obj
        end

        def self.hits_metric(client, service)
          metrics = client.list_metrics service.fetch('id')
          if metrics.respond_to?(:has_key?) && (errors = metrics['errors'])
            raise "Service metrics not read: #{errors}"
          end

          hits_metric_obj = metrics.find { |metric| metric['system_name'] == 'hits' }
          raise "Missing hits metric in service #{service.fetch('id')}" if hits_metric_obj.nil?

          hits_metric_obj
        end

        def self.create_application_plan_limit(client, service, plan, metric_id)
          # Very high limit: 4294967295 / 3600 => 1.2 M req/second during one hour to go over limit
          limit_params = { 'period' => 'hour', 'value' => 2**32 - 1 }
          limit_obj = client.create_application_plan_limit(
            plan.fetch('id'), metric_id, limit_params
          )
          if (errors = limit_obj['errors'])
            raise "Limit has not been created: #{errors}"
          end

          limit_obj
        end

        def self.delete_mapping_rules(client, service)
          mapping_rules = client.list_mapping_rules service.fetch('id')
          if mapping_rules.respond_to?(:has_key?) && (errors = mapping_rules['errors'])
            raise "Service mapping rules not read: #{errors}"
          end

          mapping_rules.each do |mapping_rule|
            client.delete_mapping_rule service.fetch('id'), mapping_rule.fetch('id')
          end
        end

        def self.create_mapping_rule(client, service, path)
          hits_metric_obj = hits_metric(client, service)
          mapping_rule_params = {
            'metric_id' => hits_metric_obj.fetch('id'), 'pattern' => path,
            'http_method' => 'GET',
            'delta' => 1
          }

          mapping_rule_obj = client.create_mapping_rule service.fetch('id'), mapping_rule_params
          if (errors = mapping_rule_obj['errors'])
            raise "MappingRule has not been created: #{errors}"
          end

          mapping_rule_obj
        end

        def self.create_application(client, plan, account)
          app_params = {
            'name' => "app_#{random_lowercase_name}",
            'description' => "app #{random_lowercase_name}"
          }

          app_obj = client.create_application(account.fetch('id'), app_params, plan_id: plan.fetch('id'))
          if (errors = app_obj['errors'])
            raise "Application has not been created: #{errors}"
          end

          app_obj
        end

        def self.account(client)
          accounts = client.list_accounts
          if accounts.respond_to?(:has_key?) && (errors = accounts['errors'])
            raise "Accounts not read: #{errors}"
          end

          raise 'No accounts available' if accounts.length.zero?

          accounts[0]
        end

        def self.create_account(client)
          account_name = "account_#{random_lowercase_name}"
          account_obj = client.signup(name: account_name, username: account_name)
          if account_obj.respond_to?(:has_key?) && (errors = account_obj['errors'])
            raise "Account not created: #{errors}"
          end

          account_obj
        end

        def self.create_backend(client, private_base_url)
          attrs = {
            name: random_lowercase_name,
            private_endpoint: private_base_url,
          }

          backend_obj = client.create_backend(attrs)
          if backend_obj.respond_to?(:has_key?) && (errors = backend_obj['errors'])
            raise "Backend not created: #{errors}"
          end

          backend_obj
        end

        def self.create_backend_usage(client, product, backend, path)
          attrs = {
            backend_api_id: backend.fetch('id'),
            path: path
          }

          backend_usage_obj = client.create_backend_usage(product.fetch('id'), attrs)
          if backend_usage_obj.respond_to?(:has_key?) && (errors = backend_usage_obj['errors'])
            raise "Backend usage not created: #{errors}"
          end

          backend_usage_obj
        end

        def self.update_service_proxy(client, service, public_base_url)
          proxy = { 'endpoint' => public_base_url }

          proxy.compact!

          unless proxy.empty?
            new_proxy_attrs = client.update_proxy service.fetch('id'), proxy

            if (errors = new_proxy_attrs['errors'])
              raise "Service proxy not updated: #{errors}"
            end

            new_proxy_attrs
          end
        end

        def self.bump_proxy_conf(client, service)
          client.proxy_deploy service.fetch('id')
        end

        def self.promote_proxy_conf(client, service)
          sandbox_proxy_cfg = client.proxy_config_latest(service.fetch('id'), 'sandbox')
          if (errors = sandbox_proxy_cfg['errors'])
            raise "Sandbox Proxy config not read: #{errors}"
          end

          res = client.promote_proxy_config(
            service.fetch('id'),
            'sandbox',
            sandbox_proxy_cfg.fetch('version'),
            'production'
          )
          if (errors = res['errors'])
            raise "Proxy not promoted: #{errors}"
          end

          res
        end

        def self.backend_hits_metric(client, backend)
          metrics = client.list_backend_metrics backend.fetch('id')
          if metrics.respond_to?(:has_key?) && (errors = metrics['errors'])
            raise "Backend metrics not read: #{errors}"
          end

          hits_metric_obj = metrics.find { |metric| metric['system_name'].include? 'hits' }
          raise "Missing hits metric in backend #{backend.fetch('id')}" if hits_metric_obj.nil?

          hits_metric_obj
        end

        def self.create_backend_mapping_rule(client, backend, method, path)
          mapping_rule_params = {
            'metric_id' => method.fetch('id'), 'pattern' => path,
            'http_method' => 'GET',
            'delta' => 1
          }

          mapping_rule_obj = client.create_backend_mapping_rule backend.fetch('id'), mapping_rule_params
          if (errors = mapping_rule_obj['errors'])
            raise "Backend MappingRule has not been created: #{errors}"
          end

          mapping_rule_obj
        end

        def self.create_backend_method(client, backend)
          hits_metric_obj = backend_hits_metric(client, backend)
          attrs = {
            'system_name' => random_lowercase_name,
            'friendly_name' => random_lowercase_name,
            'description' => random_lowercase_name
          }
          method_obj = client.create_backend_method(backend.fetch('id'), hits_metric_obj.fetch('id'), attrs)

          if (errors = method_obj['errors'])
            raise "Method has not been created: #{errors}"
          end

          method_obj
        end

        def self.update_private_endpoint(client, service, private_base_url)
          proxy = { api_backend: private_base_url }
          new_proxy_attrs = client.update_proxy service.fetch('id'), proxy

          if (errors = new_proxy_attrs['errors'])
            raise "Service proxy not updated: #{errors}"
          end

          new_proxy_attrs
        end

        # wait tries a block of code until it returns true, or the timeout is reached.
        # timeout give an upper limit to the amount of time this method will run
        # Some intervals may be missed if the block takes too long or the time window is too short.
        def self.wait(interval = 1.5, timeout = 30)
          raise 'wait expects block' unless block_given?

          end_time = Time.now + timeout
          until Time.now > end_time
            result = yield
            return if result == true

            sleep interval
          end

          raise "timed out after #{timeout} seconds"
        end

        def self.backends(client)
          b_list = client.list_backends
          if b_list.respond_to?(:has_key?) && (errors = b_list['errors'])
            raise "Backend list not read: #{errors}"
          end

          b_list
        end

        def self.backend_methods(client, backend)
          hits_metric_obj = backend_hits_metric(client, backend)
          m_list = client.list_backend_methods(backend.fetch('id'), hits_metric_obj.fetch('id'))
          if m_list.respond_to?(:has_key?) && (errors = m_list['errors'])
            raise "Backend list not read: #{errors}"
          end

          m_list
        end
      end
    end
  end
end
